### PR TITLE
Switch to github.com/golang-jwt/jwt

### PR DESCRIPTION
**Pull request by vovinacci**
Closes https://github.com/go-kit/kit/issues/1026

This PR switches to `github.com/golang-jwt/jwt` to address security vulnerability. More details in the issue and https://github.com/advisories/GHSA-w73w-5m7g-f7qc

### CODE REVIEW

**Comment by peterbourgon** [comment]
@sagikazarmark Look good to you?

**Review by sagikazarmark — APPROVED** [review]

**Comment by sagikazarmark** [comment]
After taking a more thorough look at the implementation, I'm not _sure_ that there is a breaking change, but I _suspect_ there might be.

Basically, those different types referenced in the public API are all interfaces and I assume the new package didn't break those overnight, so in _theory_ the replaced types could be backwards compatible with the old library.

In practice, though, some of the types used internally (eg. standard claims struct) are not interfaces, so those could break.

A simple test to check if there is a breaking change would be (temporarily) changing back the packages in tests to the old module to see if it works with the new code.

I'm not particularly worried about a breaking change (the library is basically abandoned after all), but if there is a breaking change, it should be released in a minor version, not a patch. Go 1.17 is just around the corner and Chris wants to cut a new release after that (hoping that we can further reduce the amount of dependency downloads).

**Comment by peterbourgon** [comment]
OK let's bench this until the next release.

**Comment by vovinacci** [comment]
@peterbourgon Do you have any preliminary release date in mind?

**Comment by AliLogic** [comment]
Any updates on this?

File: go.mod [hunk 1]
```diff
@@ -12,13 +12,13 @@ require (
...
 	github.com/clbanning/x2j v0.0.0-20191024224557-825249438eec // indirect
 	github.com/codahale/hdrhistogram v0.0.0-20161010025455-3a0bb77429bd // indirect
-	github.com/dgrijalva/jwt-go v3.2.0+incompatible
 	github.com/edsrzf/mmap-go v1.0.0 // indirect
 	github.com/franela/goblin v0.0.0-20200105215937-c9ffbefa60db // indirect
 	github.com/franela/goreq v0.0.0-20171204163338-bcd34c9993f8 // indirect
 	github.com/go-logfmt/logfmt v0.5.0
 	github.com/go-stack/stack v1.8.0
 	github.com/go-zookeeper/zk v1.0.2
+	github.com/golang-jwt/jwt v3.2.1+incompatible
```
Comment (sagikazarmark) [inline go.mod:21]: Looks like v4 is released with proper Go module support

**Review by sagikazarmark — CHANGES_REQUESTED** [review]
Can you please upgrade the library to v4?

**Comment by vovinacci** [comment]
@sagikazarmark Done

**Comment by shogo82148** [comment]
I recommend to upgrade to Go 1.17 pruned module graphs.

https://golang.org/doc/go1.17#go-command

```
go mod tidy -go=1.17
```

It enables some optimizations, such as [Lazy module loading](https://golang.org/ref/mod#lazy-loading).

**Comment by eli-darkly** [comment]
@shogo82148 If that would require setting the minimum version to 1.17 in go.mod, I would very strongly prefer that they _not_ make that change, at least not yet. Go 1.17 was just released and we can't all immediately update our projects to require it.

**Comment by sagikazarmark** [comment]
> I recommend to upgrade to Go 1.17 pruned module graphs.

This is already in progress on another branch. Not sure which will be merged first yet.

**Review by sagikazarmark — APPROVED** [review]

**Comment by andig** [comment]
> @shogo82148 If that would require setting the minimum version to 1.17 in go.mod, I would very strongly prefer that they _not_ make that change, at least not yet. Go 1.17 was just released and we can't all immediately update our projects to require it.

Imho this does not raise the minimum version.

**Comment by eli-darkly** [comment]
@andig:
> Imho this does not raise the minimum version.

Could you say more about why you think it wouldn't? My reading of the linked documentation is that pruned module graphs are only a thing in `go.mod` files that specify `go 1.17`.

**Comment by shogo82148** [comment]
pruned module graphs are just `required` directives.
So, Go 1.16 or earlier parses them as extra dependencies.
This has no negative effects, because these extra dependencies are also in `go.sum` and Go 1.16 loads whole dependencies in `go.sum`.

The `go 1.17` directive enables new language features, and changes the `go` command behavior. https://golang.org/ref/mod#go-mod-file-go
However it doesn't break backward compatibility.

When Go 2.0 comes out in the future, Go 2.0 will not be able to import modules that contain the `go 1.17` directive.

**Pull request merged**